packages/trackerless-network/src/logic/neighbor-discovery/Handshaker.ts

```diff
@@ -0,0 +1,181 @@
+        if (targetNeighbor) {
+            const accepted = await this.handshakeWithTarget(this.createRemoteHandshaker(targetNeighbor.getPeerDescriptor()))
+            if (!accepted) {
+                excludedIds.push(keyFromPeerDescriptor(targetNeighbor.getPeerDescriptor()))
+            }
+        }
+        return excludedIds
+    }
+
+    private async handshakeWithTarget(targetNeighbor: RemoteHandshaker, concurrentStringId?: string): Promise<boolean> {
+        const targetStringId = keyFromPeerDescriptor(targetNeighbor.getPeerDescriptor())
+        this.ongoingHandshakes.add(targetStringId)
+        const result = await targetNeighbor.handshake(
+            this.config.ownPeerDescriptor,
+            this.config.targetNeighbors.getStringIds(),
+            concurrentStringId
+        )
+        if (result.accepted) {
+            this.config.targetNeighbors.add(this.createRemoteNode(targetNeighbor.getPeerDescriptor()))
+            this.config.connectionLocker.lockConnection(targetNeighbor.getPeerDescriptor(), this.config.randomGraphId)
+        }
+        if (result.interleaveTarget) {
+            await this.handshakeWithInterleaving(result.interleaveTarget, targetStringId)
+        }
+        this.ongoingHandshakes.delete(targetStringId)
+        return result.accepted
+    }
+
+    private async handshakeWithInterleaving(target: PeerDescriptor, interleavingFrom: string): Promise<boolean> {
+        const targetNeighbor = new RemoteHandshaker(
+            target,
+            this.config.randomGraphId,
+            this.client
+        )
+        const targetStringId = keyFromPeerDescriptor(targetNeighbor.getPeerDescriptor())
+        this.ongoingHandshakes.add(targetStringId)
+        const result = await targetNeighbor.handshake(
+            this.config.ownPeerDescriptor,
+            this.config.targetNeighbors.getStringIds(),
+            undefined,
+            interleavingFrom
+        )
+        if (result.accepted) {
+            this.config.targetNeighbors.add(this.createRemoteNode(targetNeighbor.getPeerDescriptor()))
+            this.config.connectionLocker.lockConnection(targetNeighbor.getPeerDescriptor(), this.config.randomGraphId)
+        }
+        this.ongoingHandshakes.delete(targetStringId)
+        return result.accepted
+    }
+
+    private createRemoteHandshaker(targetPeerDescriptor: PeerDescriptor): RemoteHandshaker {
+        return new RemoteHandshaker(targetPeerDescriptor, this.config.randomGraphId, this.client)
+    }
+
+    private createRemoteNode(targetPeerDescriptor: PeerDescriptor): RemoteRandomGraphNode {
+        return new RemoteRandomGraphNode(
+            targetPeerDescriptor,
+            this.config.randomGraphId,
+            toProtoRpcClient(new NetworkRpcClient(this.config.rpcCommunicator!.getRpcClientTransport()))
+        )
+    }
+
+    public getOngoingHandshakes(): Set<string> {
```

Comment:
Is this info needed only in tests? (via `RandomGraphNode#getNumberOfOutgoingHandshakes`)

Maybe could remove this method and RandomGraphNode#getNumberOfOutgoingHandshakes if not needed outside tests? Or simplify to be something like `hasOngoingHandshakes`, as that is what the tests query.